### PR TITLE
MC: Multirotor mixer rewrite & Airmode for Yaw

### PR DESCRIPTION
**Pull request by bkueng**
For acrobatic flights airmode for yaw is a requirement. But instead of extending the existing mixer, I joined forces with @bresch. He came up with a nice generic approach to handle saturations, making it easier to define different axis prioritizations.
It is described here: https://github.com/Auterion/Flight_Control_Prototyping_Scripts/tree/master/control_allocation. It is also a step towards 6D control allocation.

This PR does:
- complete rewrite of the MC mixer according to the described method above
- extends `MC_AIRMODE` by a third mode that allows yaw boosting
- enforces the use of an arming switch if yaw airmode is selected (stick arming is just too dangerous if used with yaw airmode, I'm sure nobody wants that)
- Stabilized: allow yaw control at 0 throttle if yaw airmode is enabled
- add unit-tests for the mixer (the existing ones don't check the actual mixing algorithms), see below

### Differences
The behavior for both existing Airmode settings (disabled & roll/pitch) is exactly the same (based on running through different test-cases), except for:
- The new algorithm handles asymmetric frames correctly
- While comparing the algorithms, I found a case where the existing mixer would reduce the demanded yaw more than actually required (and it is not a very extraordinary case). The new algorithm does it correctly.
- very high roll/pitch demands at low throttle are handled a bit differently by the new algorithm (no surprising behavior though).

CPU load increases by 0.5% for the airmode=disabled case.

### Testing level
This has run through several levels of testing:
- prototyping and corner-case testing in python
- extensive HIL simulations for all 3 modes
- actual flight testing:
  - MC_AIRMODE=0
    - https://logs.px4.io/plot_app?log=c0d68572-5763-4253-87a6-76a30cc25a2a
    - https://logs.px4.io/plot_app?log=5156eed6-9593-40f8-af07-c48a757dcd29
      In case you are wondering about the flips in Stabilized: I set the minimum throttle to 0, and whenever I cut throttle, the vehicle would not stabilize anymore as expected. When increasing the throttle, it immediately recovers from any attitude and any angular velocity (which is nice to see. Kudos to @MaEtUgR for the extremely robust attitude controller).
    - https://logs.px4.io/plot_app?log=84681a24-3ac7-4db7-b1d9-05e71a326cd7
  - MC_AIRMODE=1
    - https://logs.px4.io/plot_app?log=d1eb9c03-022d-41e5-8f4d-2389fb0a5884
    - https://logs.px4.io/plot_app?log=b9a76438-c05c-4c23-94d5-eea51c7c2d53
  - MC_AIRMODE=2
    - https://logs.px4.io/plot_app?log=291fc0df-cf51-42bb-8a35-e9396ecb104b
    - https://logs.px4.io/plot_app?log=cc7bde3e-09df-4624-a93d-fb1ba07aed96
    - https://logs.px4.io/plot_app?log=4452abb0-5dd9-4453-985c-18f93e676384
    - https://logs.px4.io/plot_app?log=1d28d82c-bfe3-4ddc-8a6e-2458ae2ae39e
    - https://logs.px4.io/plot_app?log=7266e561-a286-4c0d-b96f-ccbb3ee8d972
- unit-tests

#### Unit tests
I added the python script used to prototype to also serve as basis for unit-tests: it provides ground-truth data that is then tested against the C++ implementation. This allows to:
- refactor the existing C++ implementation with high confidence that the code does the same thing
- quickly prototype different saturation strategies or completely different algorithms
- test the output of specific (corner) cases
- ensures the python implementation stays in sync with the C++ implementation

To run them:
```
cd src/lib/mixer
make tests
```
It will run though all 3 airmodes with different geometries and actuator control commands (almost 1000 test cases all in all).
I also let it run against the previous mixer implementation which showed the exact differences.
It builds a separate binary at the moment beause it does not really fit into the existing unit-testing framework (requires calling the python script), helps minimizing dependencies and it allows a very fast testing workflow. It could however be a bit more integrated.

### TODO
- Testing on a board with an IO (and look at its CPU load)
- Testing some other geometries (I tested quad x & hex x) (e.g. asymmetric ones)

### CODE REVIEW

File: src/modules/commander/Commander.hpp [hunk 1]
```diff
@@ -110,7 +110,10 @@ class Commander : public ModuleBase<Commander>, public ModuleParams
 		(ParamInt<px4::params::COM_POS_FS_GAIN>) _failsafe_pos_gain,
 
 		(ParamInt<px4::params::COM_LOW_BAT_ACT>) _low_bat_action,
-		(ParamFloat<px4::params::COM_DISARM_LAND>) _disarm_when_landed_timeout
+		(ParamFloat<px4::params::COM_DISARM_LAND>) _disarm_when_landed_timeout,
+
+		(ParamInt<px4::params::MC_AIRMODE>) _airmode,
```
Comment (dagar) [inline src/modules/commander/Commander.hpp]: Could you use the old param API or do something else here? I'd like to preserve the ability to do FW or MC only builds. I'll add test configs to the CI system.

**Comment by hamishwillee** [comment]
@dagar @bresch Doesn't this have an impact on https://docs.px4.io/en/config_mc/pid_tuning_guide_multicopter.html (and to a lesser extent https://docs.px4.io/en/config_mc/racer_setup.html) ?

File: src/modules/commander/Commander.hpp (hunk 1 above)
Comment (bkueng) [reply src/modules/commander/Commander.hpp]: Done

**Comment by bkueng** [comment]
> @dagar @bresch Doesn't this have an impact on https://docs.px4.io/en/config_mc/pid_tuning_guide_multicopter.html (and to a lesser extent https://docs.px4.io/en/config_mc/racer_setup.html) ?

Only slightly: we can mention that there are separate airmodes for roll/pitch only and roll/pitch/yaw.

**Comment by MaEtUgR** [comment]
Woooah, awsome!! Big thanks to @bkueng and @bresch 👍

**Comment by bkueng** [comment]
I checked the CPU impact on the IO:
Loop rate in Hz for this branch:
- MC_AIRMODE=0: 765 Hz
- MC_AIRMODE=1: 945 Hz
- MC_AIRMODE=2: 1240 Hz

master:
- MC_AIRMODE=0: 1140 Hz
- MC_AIRMODE=1: 1160 Hz

---

@PX4/testflights can you test this please as following?
- test on different boards (with/without IO)
- test on different multirotors (preferably hex/octo and asymmetric if you have), and VTOLs
- First test with Airmode disabled
- Start off **carefully**, until you're confident that the vehicle behaves as expected. Then go crazy, pushing the vehicle to the limits, to hit the mixer saturation corner cases.
- test different airmode settings: `MC_AIRMODE` enabled/disabled.

File: src/lib/mixer/mixer_multirotor.cpp [hunk 2]
```diff
@@ -146,157 +188,202 @@ MultirotorMixer::from_text(Mixer::ControlCallback control_cb, uintptr_t cb_handl
...
+
+	// Do full mixing
+	for (unsigned i = 0; i < _rotor_count; i++) {
+		outputs[i] = roll * _rotors[i].roll_scale +
+			     pitch * _rotors[i].pitch_scale +
+			     yaw * _rotors[i].yaw_scale +
+			     thrust * _rotors[i].thrust_scale;
+
+		// Thrust will be used to unsaturate if needed
+		_tmp_array[i] = _rotors[i].thrust_scale;
```
Comment (Stifael) [inline src/lib/mixer/mixer_multirotor.cpp]: I am trying to understand the new mixing method. 
For `mix_airmode_rpy`, only thrust will be used to unsaturate. I don't really get why thrust_scale is delta_output?

Comment (Stifael) [reply src/lib/mixer/mixer_multirotor.cpp]: Or put it differently: what exactly should delta_output represent?

File: src/lib/mixer/mixer_multirotor.cpp [hunk 3]
```diff
@@ -146,157 +188,202 @@ MultirotorMixer::from_text(Mixer::ControlCallback control_cb, uintptr_t cb_handl
...
+			if (k < k_min) { k_min = k; }
+
+			if (k > k_max) { k_max = k; }
+
+			sat_status.flags.motor_pos = true;
 		}
+	}
+
+	// Reduce the saturation as much as possible
+	return k_min + k_max;
```
[the remark this answers is not in the document]
Comment (Stifael) [reply src/lib/mixer/mixer_multirotor.cpp]: never mind, got it

File: src/lib/mixer/mixer_multirotor.cpp (hunk 2 above)
Comment (Stifael) [reply src/lib/mixer/mixer_multirotor.cpp]: never mind got it. But the name is somewhat confusing, given that it is not a delta

Comment (bkueng) [reply src/lib/mixer/mixer_multirotor.cpp]: What name would you suggest? Would `desaturation_vector` work?

File: src/lib/mixer/mixer_multirotor.cpp [hunk 4]
```diff
@@ -146,157 +188,202 @@ MultirotorMixer::from_text(Mixer::ControlCallback control_cb, uintptr_t cb_handl
...
-		/*
-			implement simple model for static relationship between applied motor pwm and motor thrust
-			model: thrust = (1 - _thrust_factor) * PWM + _thrust_factor * PWM^2
-			this model assumes normalized input / output in the range [0,1] so this is the right place
-			to do it as at this stage the outputs are in that range.
-		 */
+		_tmp_array[i] = _rotors[i].thrust_scale;
+	}
+
+	minimize_saturation(_tmp_array, outputs, _saturation_status, -1000.f);
```
Comment (bresch) [inline src/lib/mixer/mixer_multirotor.cpp]: I think we should also use the "reduce only" flag here:
```
// only reduce thrust
minimize_saturation(_tmp_array, outputs, _saturation_status, 0.f, 1.f, true);
```

File: src/lib/mixer/mixer_multirotor.py [hunk 5]
```diff
@@ -0,0 +1,380 @@
...
+    m_sp_yaw_only = np.matlib.zeros(m_sp.size).T
+    m_sp_yaw_only[2, 0] = m_sp[2, 0]
+    u_p = u + P * m_sp_yaw_only
+
+    # Change yaw acceleration to unsaturate the outputs if needed (do not change roll/pitch),
+    # and allow some yaw response at maximum thrust
+    u_r_dot = P[:,2]
+    u_pp = minimize_sat(u_p, u_min, u_max+0.15, u_r_dot)
+    u_T = P[:, 3]
+    u_ppp = minimize_sat(u_pp, -1000, u_max, u_T)
```
Comment (bresch) [inline src/lib/mixer/mixer_multirotor.py]: use `u_min` instead of -1000. Should we add the flag `reduce only` like in the cpp code?

File: src/lib/mixer/mixer_multirotor.cpp (hunk 2 above)
Comment (Stifael) [reply src/lib/mixer/mixer_multirotor.cpp]: yep, thats sounds better. `delta` sounds like the difference between two values

File: src/lib/mixer/mixer.h [hunk 6]
```diff
@@ -614,6 +641,20 @@ class __EXPORT MultirotorMixer : public Mixer
...
 			float yaw_scale,
 			float idle_speed);
+
+	/**
+	 * Constructor (for testing).
+	 *
+	 * @param control_cb		Callback invoked to read inputs.
+	 * @param cb_handle		Passed to control_cb.
+	 * @param rotors		control allocation matrix
+	 * @param rotor_count		length of rotors
```
[the remark this answers is not in the document]
Comment (bkueng) [reply src/lib/mixer/mixer.h]: That too, but it refers to the length of the `rotors` array. I updated it.

File: src/lib/mixer/mixer_multirotor.cpp (hunk 2 above)
Comment (bkueng) [reply src/lib/mixer/mixer_multirotor.cpp]: done

**Comment by bkueng** [comment]
@PX4/testflights please hold off with testing one more day so that I can re-test @bresch's changes first.

**Comment by bkueng** [comment]
@bresch I pushed the changes to use the `reduce_only` flag for yaw as well, it works as expected in HIL.

@PX4/testflights you can now go ahead and test it.

**Comment by Avysuarez** [comment]
Tested on pixhawk 4 mini. 
Air mode disabled.
https://review.px4.io/plot_app?log=0b9b50a5-24ba-40e3-879b-bececc61c96c
Air mode enabled.
https://review.px4.io/plot_app?log=71a93818-639c-47d2-9e9a-252b07f11f67

**Comment by dannyfpv** [comment]
tested on pixhawk v4 pro
MC_AIRMODE=disabled 
https://review.px4.io/plot_app?log=3755c03c-99fb-4474-902f-b67c55801811

MC_AIRMODE=enable
https://review.px4.io/plot_app?log=23c57450-3a32-4af2-b024-45556bd2d768

**Comment by bkueng** [comment]
Yay, flips with the F450, nice :)

The logs look good, I went through all of them.
Let's merge it then.

**Comment by mrpollo** [comment]
:shipit:
